ext/openssl/ossl_ts.c

```diff
@@ -0,0 +1,1492 @@
+ * a given Request. Mandatory parameters are the Request associated to this
+ * Response, and an OpenSSL::X509::Store of trusted roots.
+ *
+ * Intermediate certificates can optionally be supplied for creating the
+ * certificate chain. These intermediate certificates must all be
+ * instances of OpenSSL::X509::Certificate.
+ *
+ * If validation fails, several kinds of exceptions can be raised:
+ * * TypeError if types don't fit
+ * * TimestampError if something is wrong with the timestamp token itself, if
+ *   it is not conformant to the Request, or if validation of the timestamp
+ *   certificate chain fails.
+ *
+ * call-seq:
+ *       response.verify(Request, root_store) -> Response
+ *       response.verify(Request, root_store, [intermediate_cert]) -> Response
+ */
+static VALUE
+ossl_ts_resp_verify(int argc, VALUE *argv, VALUE self)
+{
+    VALUE ts_req, store, intermediates;
+    TS_RESP *resp;
+    TS_REQ *req;
+    X509_STORE *x509st;
+    TS_VERIFY_CTX *ctx;
+    STACK_OF(X509) *x509inter = NULL;
+    PKCS7* p7;
+    X509 *cert;
+    int status, i, ok;
+
+    rb_scan_args(argc, argv, "21", &ts_req, &store, &intermediates);
+
+    GetTSResponse(self, resp);
+    GetTSRequest(ts_req, req);
+    x509st = GetX509StorePtr(store);
+
+    if (!(ctx = TS_REQ_to_TS_VERIFY_CTX(req, NULL))) {
+	ossl_raise(eTimestampError, "Error when creating the verification context.");
+    }
+
+    if (intermediates != Qnil) {
+	x509inter = ossl_protect_x509_ary2sk(intermediates, &status);
+	if (status) {
+	    TS_VERIFY_CTX_free(ctx);
+	    rb_jump_tag(status);
+	}
+    } else {
+	x509inter = sk_X509_new_null();
+    }
+
+    if (!(p7 = TS_RESP_get_token(resp))) {
+	TS_VERIFY_CTX_free(ctx);
+	sk_X509_pop_free(x509inter, X509_free);
+	ossl_raise(eTimestampError, "TS_RESP_get_token");
+    }
+    for (i=0; i < sk_X509_num(p7->d.sign->cert); i++) {
+	cert = sk_X509_value(p7->d.sign->cert, i);
+	X509_up_ref(cert);
+	sk_X509_push(x509inter, cert);
+    }
+
+    TS_VERIFY_CTS_set_certs(ctx, x509inter);
+    TS_VERIFY_CTX_add_flags(ctx, TS_VFY_SIGNATURE);
+    TS_VERIFY_CTX_set_store(ctx, x509st);
+
+    ok = TS_RESP_verify_response(ctx, resp);
+
+    /* WORKAROUND:
+     *   X509_STORE can count references, but X509_STORE_free() doesn't check
+     *   this. To prevent our X509_STORE from being freed with our
+     *   TS_VERIFY_CTX we set the store to NULL first.
+     *   Fixed in OpenSSL 1.0.2; bff9ce4db38b (master), 5b4b9ce976fc (1.0.2)
+     */
+    TS_VERIFY_CTX_set_store(ctx, NULL);
+    TS_VERIFY_CTX_free(ctx);
+
+    if (!ok)
+	ossl_raise(eTimestampError, "TS_RESP_verify_response");
+
+    return self;
+}
+
+void
+ossl_ts_verify_ctx_free(TS_VERIFY_CTX *ctx)
+{
+    
+}
+
+static ASN1_INTEGER *
+ossl_tsfac_serial_cb(struct TS_resp_ctx *ctx, void *data)
+{
+    return (ASN1_INTEGER *)data;
+}
+
+static int
+ossl_tsfac_time_cb(struct TS_resp_ctx *ctx, void *data, long *sec, long *usec)
+{
+    *sec = *((long *)data);
+    *usec = 0;
+    return 1;
+}
```

Comment:
Please move these two functions to below the definitions of ossl_ts_token_* functions.